From 0a5a1b8b4acb254f71bc8d74e1cb7a583aba7c62 Mon Sep 17 00:00:00 2001
From: Hans de Goede <hansg@kernel.org>
Date: Sat, 30 Aug 2025 14:31:11 +0200
Subject: [PATCH 1/3] Fix BotController->toCover_ NULL pointer deref crash

In 2017 there where report with the Fedora packages of marsshooter crashing
due to BotController->toCover_ being NULL. Fix this by storing the radius
to cover as float in BotController->toCoverRadius_, avoiding a NULL
pointer deref if this gets deferred before it is set.
---
 include/Controllers/BotController.hpp        | 2 +-
 src/Controllers/BotController.cpp            | 2 +-
 src/Controllers/BotControllerTeamActions.cpp | 5 +++--
 3 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/include/Controllers/BotController.hpp b/include/Controllers/BotController.hpp
index 8947f13..29df778 100644
--- a/include/Controllers/BotController.hpp
+++ b/include/Controllers/BotController.hpp
@@ -99,7 +99,7 @@ class BotController: public Controller {
         void     shootPoint(Vector2f const& location, bool avoidTeamMembers = true);
 
         Vector2f nextRoutePoint_;
-        TacticalZone* toCover_;
+        float toCoverRadius_;
         const float strength_;
 
         //path variables
diff --git a/src/Controllers/BotController.cpp b/src/Controllers/BotController.cpp
index eb3e82a..53cafcc 100644
--- a/src/Controllers/BotController.cpp
+++ b/src/Controllers/BotController.cpp
@@ -38,7 +38,7 @@ BotController::BotController(Player* slave, float strength):
     specialChangeTimer_(randomizer::random(0.f, 0.5f)),
     currentJob_(Job::jCharge, 1),
     nextRoutePoint_(FLT_MAX, FLT_MAX),
-    toCover_(NULL),
+    toCoverRadius_(0.f),
     strength_(strength) {}
 
 void BotController::update() {
diff --git a/src/Controllers/BotControllerTeamActions.cpp b/src/Controllers/BotControllerTeamActions.cpp
index fbaf0c0..34d0f88 100644
--- a/src/Controllers/BotControllerTeamActions.cpp
+++ b/src/Controllers/BotControllerTeamActions.cpp
@@ -175,10 +175,11 @@ void  BotController::attackTarget() {
 
 void BotController::protectZone() {
     if (nextRoutePoint_.x_ == FLT_MAX) {
-        toCover_ = static_cast<TacticalZone*>(currentJob_.object_);
+        TacticalZone* toCover_ = static_cast<TacticalZone*>(currentJob_.object_);
         nextRoutePoint_ = toCover_->getRandomPoint();
+        toCoverRadius_ = toCover_->radius_;
     }
-    if (moveTo(nextRoutePoint_, 0.5f, false, toCover_->radius_ / 4.f + 10.f)) {
+    if (moveTo(nextRoutePoint_, 0.5f, false, toCoverRadius_ / 4.f + 10.f)) {
         nextRoutePoint_.x_ = FLT_MAX;
     }
 }
-- 
2.51.0

